Simplify minterm generation in Regex NonBacktracking engine

olsaarik opened the pull request:
PartitionTree had a holdover from when it was tracking the predicates that made up each minterm, which caused it to use order n too much memory for some patterns: nodes with the same predicate as the parent are no longer created. This also significantly simplified the algorithm.

The effect is particularly evident in patterns with lots of non-overlapping predicates. For the pattern "qwertyuuiopasdfghjklzxcvbnm" this dropped the number of PartitionTree nodes created from 378 to 53.

olsaarik committed:
Simplify and optimize minterm generation

PartitionTree had a holdover from when it was tracking the predicates
that made up each minterm, which caused it to use order n too much
memory for some patterns.
Nodes with the same predicate as the parent are no longer created.
This also significantly simplified the algorithm.

stephentoub reviewed (approved):
Less code and cheaper. Yay :)

olsaarik committed:
Make MintermGenerator take an IEnumerable

Also remove the deduplication in there, since its now optional and the
only call site already passes in a unique set of predicates.
